### PR TITLE
Store TempDir for the lifetime of a program

### PR DESCRIPTION
**Pull request by blackbeam**
This solves #34.

Might be an overkill but i can't find another solution.
I've tried to store tempdir within Pkcs12 and even tried to store it in a wrapper for SecKeychain but with no luck.

### CODE REVIEW

File: src/imp/security_framework.rs [hunk 1]
```diff
@@ -14,9 +15,16 @@ use self::tempdir::TempDir;
...
 use std::io;
 use std::error;
+use std::sync::Mutex;
+use std::sync::atomic::{AtomicBool, Ordering};
 
 use Protocol;
 
+lazy_static! {
+    static ref AT_EXIT_SET: AtomicBool = AtomicBool::new(false);
+    static ref TEMP_DIRS: Mutex<Vec<TempDir>> = Mutex::new(vec![]);
```
Comment (sfackler) [inline src/imp/security_framework.rs]: Can we treat this as a map from the raw DER data to the directory? That way we don't end up with tons of directories when the same identity is imported multiple times.

File: src/imp/security_framework.rs [hunk 2]
```diff
@@ -101,6 +110,16 @@ impl Pkcs12 {
         // FIXME: Compare the certificates for equality using CFEqual
         let identity_cert = try!(import.identity.certificate()).to_der();
 
+        if !AT_EXIT_SET.swap(true, Ordering::Relaxed) {
```
Comment (sfackler) [inline src/imp/security_framework.rs]: You can use std::sync::Once to do this.

**Comment by blackbeam** [comment]
If SecKeystore isn't cached, then you'll get "keychain with this name already exists" error, so now it's also cached.

File: src/imp/security_framework.rs [hunk 3]
```diff
@@ -79,17 +94,27 @@ pub struct Pkcs12 {
...
+            Occupied(entry) => entry.get().0.clone(),
         };
 
-        let mut keychain = try!(keychain::CreateOptions::new().password(pass).create(
-            dir.path().join("tmp.keychain"),
-        ));
-        // disable lock on sleep and timeouts
-        try!(keychain.set_settings(&KeychainSettings::new()));
-
         let mut imports = try!(
```
Comment (sfackler) [inline src/imp/security_framework.rs]: Will things behave okay if we import the same identity into the same keychain twice?

**Comment by blackbeam** [comment]
>Why do we need a custom hash function here?

It's faster for long keys. It's just a habit and i'll remove it if you say so.

> Will things behave okay if we import the same identity into the same keychain twice?

I've tried to import same der multiple times and it looks like the number of identities in a keychain stays the same (1 identity as shown by ItemSearchOptions).
Anyway there is no need to import same der multiple times so cache now simply stores `Pkcs12` instance.

**Comment by sfackler** [comment]
I'd like to see some benchmarks indicating that the custom hasher is worth it.

The new behavior does mean the password is ignored after the first decode for a given identity.

**Comment by blackbeam** [comment]
> I'd like to see some benchmarks indicating that the custom hasher is worth it.

[this one](http://aras-p.info/blog/2016/08/02/Hash-Functions-all-the-way-down/)

>The new behavior does mean the password is ignored after the first decode for a given identity.

🤦‍♂️
Well. First of all i tried to simply import pkcs12 without specifying keychain but than i found commit message of 304b00e that states that it'll not work.
Creating another temporary keychain only for password verification looks really ugly.
I ended up with map changed from `der => (Pkcs12, TempDir)` to `(der, hash(pass)) => (Pkcs12, TempDir)` using CommonCrypto bindings.

**Comment by sfackler** [comment]
I'm talking about this use case in particular, not hash functions in general.

**Comment by blackbeam** [comment]
> I'm talking about this use case in particular, not hash functions in general.

Code:
```rust
#[bench]
fn load_pkcs12(bencher: &mut tst::test::Bencher) {
    let buf = include_bytes!("../test/identity.p12");
    bencher.iter(|| Pkcs12::from_der(buf, "mypass").unwrap());
}
```

Default hasher:
```
test test::load_pkcs12 ... bench:       2,249 ns/iter (+/- 333)
```

xxHash:
```
test test::load_pkcs12 ... bench:       1,458 ns/iter (+/- 935)
```

**Comment by sfackler** [comment]
Hmm, did you ever try using a single keychain for all of the identities? It seems like that'd simplify things quite a lot.

**Comment by blackbeam** [comment]
I switched to a single keychain and also have any of password or der cache removed. Also looks like  it is ok to call SecPKCS12Import multiple times on the same der data with the same keychain being specified, however this does not stated clearly in the docs.
I've added `import_same_identity_multiple_times` test end [run it](https://travis-ci.org/blackbeam/rust-native-tls/builds/268885305?utm_source=github_status&utm_medium=notification) on every osx version available on travis.

**Comment by sfackler** [comment]
Awesome, thanks!

**Pull request merged**